### ISSUE TITLE
feat: Allow deployment of langfuse with given API token

### ISSUE THREAD

**Issue by matthieudelaro** [langfuse/langfuse#517]
### Describe the feature you'd like to request

I would like to be able to spawn a new instance of langfuse with docker-compose / k8s.
I also want to spawn langchain services (langflow, python scripts) which are going to connect to this langfuse.

### Describe the solution you'd like to see

So I would like to provide langfuse with the API keys which it will accept connections from.

(Ideally, there would also be automatic creation of users. But maybe it's better to keep that in a separate PR.)

### Additional information

I'm open to contributing by creating a PR, if clear instructions are given about the arch of langfuse which is relevant to the issues, and locations in the source code where I should put modifications.

### Contribute

- [X] Yes, I can implement this and raise a PR

**Comment by matthieudelaro** [comment]
Idea if the database schema seems kinda stable to @marcklingen:

how about provisionning a DB for langfuse, in which we inject credential data on start?

**Comment by marcklingen** [comment]
Agree, this would be super useful!

Context
- Currently migrations are run when the container starts, see [entrypoint.sh](https://github.com/langfuse/langfuse/blob/main/entrypoint.sh) - I would caution against adding any steps that make changes to the database here
- Injecting credential data is possible, but secret keys are [stored as a hash](https://github.com/langfuse/langfuse/blob/af3e64558037546ba55b1d695577a809f9d772da/src/features/public-api/lib/apiKeys.ts#L17) (with salt), if you make changes to the database, you'd need to follow this as well
- The [seeder](https://github.com/langfuse/langfuse/blob/main/prisma/seed.ts) used in development creates a default user and project. It is executed via [npm run db:seed](https://github.com/langfuse/langfuse/blob/af3e64558037546ba55b1d695577a809f9d772da/package.json#L18) (optionally db:seed:examples). The seeder is also [used in CI](https://github.com/langfuse/langfuse/blob/af3e64558037546ba55b1d695577a809f9d772da/.github/workflows/pipeline.yml#L52) to test against a running Langfuse instance with default API keys.
- I'd not go for creating multiple users as of now but rather have a simple way to create a single one for more headless deployments where the UI might be used, but not interactively in the beginning.

Fastest route
- add additional environment to seeder, similar to `environment === "examples"`
  - configure the default users and keys to be created via env vars
     - user
     - pw
     - public key
     - secret key
- docs on how this works
  - note on the env var that is necessary
  - `git clone [repo] && npm i && npm run db:createdefaultuser`

**Comment by marcklingen** [comment]
@matthieudelaro have you made an attempt to add this? Happy to contribute or help

**Comment by alexis-alcmeon** [comment]
Hello, I've tried implement this feature in a new PR.
Don't know where I could contribute to the documentation so I haven't done it, but I'm willing to do it.

**Comment by tzsombor95** [comment]
When do you plan to submit this pull request? @marcklingen 
This would be a very useful feature for me as well.

**Comment by marcklingen** [comment]
Hi everyone, I'm super sorry for moving slow on this. I'm currently thinking about how to best achieve this. This PR is a solid approach but also quite hacky as we rely on the seeder. To make this more maintainable, provisioning projects via an API would be much better based on an org: https://github.com/langfuse/langfuse/issues/1930

**Comment by olokshyn** [comment]
Hey there! I'm coming from a project where we use Terraform to deploy to QA and PROD environments to ensure the deployments are reproducible. In addition, we have to self-host everything that contains user data.

Terraform simultaneously deploys both Langfuse and the chatbot service that tries to authenticate with Langfuse. Thus, it's not feasible to manually create a project and an API key in Langfuse, since the API key has to be known before the deployment of the chatbot service. Besides, developers are forbidden to make any infrastructure changes manually on QA or PROD.

This feature request is as close as it gets to my problem across the Langfuse documentation. Or am I missing something?

@marcklingen please advise if this automatically deployed self-hosted Langfuse use-case is supported in the current version of Langfuse, and what would be the best way to achieve that. Or should we wait for this feature request to be implemented?

**Comment by kryvokhyzha** [comment]
Hi everyone, any updates on this issue? Or can you recommend other ways to set default API keys for automatic deployment?
@marcklingen

**Comment by shreyanair-ad** [comment]
is there API or sdk support to generate the API keys?

**Comment by marcklingen** [comment]
Hi all! @matthieudelaro, @tzsombor95, @kryvokhyzha, @shreyanair-ad, @tzsombor95

I'm happy to share an update. We've just released headless initialization via the `LANGFUSE_INIT_*` environment variables. You can learn more about this change on our changelog: https://langfuse.com/changelog/2024-09-04-headless-initialization-or-self-hosted-deployments

Thank you for your patience. We wanted to add this feature much sooner, but it was blocked by a recent organizational change that could have otherwise caused the configuration to break.

Please let me know if this solves your use case and if you have any feedback.

---

Example from the Langfuse CI (GitHub Actions) where we run a langfuse instance headless via this feature to integration test all apis

https://github.com/langfuse/langfuse/blob/898cca56f994e9743630d761d5d26d0c4c77a44e/.github/workflows/pipeline.yml#L121-L132

**Comment by EricAntidot** [comment]
Hi there,
This is nice and definitely better than before but it does not completely fullfil my need where I wanted to create a bunch of organization / project / user.
I want to do that if they do not already exist.

**Comment by marcklingen** [comment]
@matthieudelaro, thanks for creating the new thread, let's centralize discussions on an admin api that goes beyond initialization via environment variables there: https://github.com/orgs/langfuse/discussions/3243
